File: packages/server/ExchangeGame/GameObjects/Exchange.cs
using Microsoft.Extensions.Logging;
using System;
using System.Collections.Generic;

namespace ExchangeGame
{
    public class Exchange: GameObject
    {
        private List<Call> _calls = new List<Call>();
        private Action<int, Exchange, Attendee> _updateScore;
        
        // TODO make configurable
        private const int FAILED_CONNECTION_SCORE = 5;

        public Exchange(string displayName, Action<int, Exchange, Attendee> updateScore): base(displayName)
        {
            _updateScore = updateScore;
           _logger.LogInformation($"Exchange {DisplayName} is in the game");
        }

        public void AddCall(Call call)
        {
            call.OnComplete += OnCallComplete;
            call.OnTimeout += OnCallTimeout;
            _calls.Add(call);
        }

        private void OnCallComplete(Call call)
        {
            _calls.Remove(call);
            _logger.LogInformation($"Exchange {DisplayName} removing call {call.DisplayName}");
        }

        private void OnCallTimeout(Call call)
        {
            _calls.Remove(call);
            _updateScore(call.Score, null, null);
        }

        public bool Connect(Attendee attendee)
        {
            var sucess = false;
            foreach(var call in _calls)
            {
                if (call.TryConnect(attendee))
                {
                    sucess = true;
                    break;
                }
            }

            if (!sucess)
            {
                _updateScore(FAILED_CONNECTION_SCORE, this, attendee);
            }

            return sucess;
        }
    }
}

File: packages/server-v2/src/GameObjects/Exchange.ts
import { generateId } from "../utils";
import Attendee from "./Attendee";
import Call from "./Call";
import GameObject from "./GameObject";

type UpdateScoreCallback =(diff: number, exchange?: Exchange, attendee?: Attendee) => void; 

export default class Exchange {
  public id: number;
  public displayName: string;
  private _calls: Call[] = [];
  private _updateScore: UpdateScoreCallback
  private FAILED_CONNECTION_SCORE = 5;

  constructor(displayName: string, updateScore: UpdateScoreCallback) {
    this.displayName = displayName;
    this.id = generateId();
    this._updateScore = updateScore;
    console.log(`Exchange ${this.displayName} is in the game`);
  }

  public addCall(call: Call) {
    call.onComplete.push(this.onCallComplete);
    call.onTimeout.push(this.onCallTimeout);
    this._calls.push(call);
  }

  public connect(attendee: Attendee): boolean {
    let sucess = false;
    for(let i=0;i<this._calls.length;i++) {
      if (this._calls[i].tryConnect(attendee)) {
        sucess = true;
        break;
      }
    }

    if (!sucess) {
      this._updateScore(this.FAILED_CONNECTION_SCORE, this, attendee);
    }

    return sucess;
  }

  private onCallComplete = (call: Call) => {
    const idx = this._calls.indexOf(call);
    if (idx !== -1) {
      this._calls.splice(idx, 1);
    }
    console.log(`Exchange ${this.displayName} removing call ${call.displayName}`)
  }

  private onCallTimeout = (call: Call) => {
    const idx = this._calls.indexOf(call);
    if (idx !== -1) {
      this._calls.splice(idx, 1);
    }
    this._calls.splice(idx, 1);
    this._updateScore(call.score);
  }
}